feat: html streaming

digitalsadhu opened the pull request:
This PR is a first attempt at an API for HTML streaming. This is made possible by the early hints work which makes it possible to get assets from podlets before the rest of the podlet content.

Comments thoughts inline..

digitalsadhu commented on lib/layout.js:
```diff
@@ -615,6 +616,20 @@ export default class PodiumLayout {
+                    this.#view(incoming, responseStream, ...args);
+                    // return the stream for the developer to work with
+                    return responseStream;
+                };
```
I went with a new function on res here. res.podiumStream which can be passed additional arguments for html template and returns a stream object to work with after setting up streaming internally.

digitalsadhu replied:
with this approach, we can stay non breaking and any page that wants to implement streaming just needs to ensure that all podlets on the page are updated to a a version of @podium/podlet that supports early hints.

digitalsadhu replied:
existing templates continue to work as before since we just inject a token and use it to split the template into 2 pieces. We then immediately send the first (header) and hand control to the developer to do streaming. Once the developer is finished and hands control back, we send the footer and terminate the stream.

trygve-lie replied:
> Worth noting that there will need to be a slightly different implementation for Fastify than Express

The difference are under the hood or also in the public API surface?

digitalsadhu replied:
The only difference would be in how we pipe the streams in the fastify plugin, api surface should be the same for the end user

digitalsadhu commented on lib/response-stream.js:
```diff
@@ -0,0 +1,37 @@
+    end() {
+        setTimeout(() => this.push(null), 0);
+    }
+}
```
This Readable stream class probably belongs in the Podium utils package. It provides a .send() method to push chunks of html onto the response, a .done() method to be used by the developer to signal when they are finished pushing content, as well as a .end() method to indicate that the app has no more html to push and closes the stream.

Essentially, html template uses .send() to push the document head, then the application developer uses .send() to push the content and calls .done() when they are finished at which point HTML template takes over again and uses .send() to push the final bits of the template and ends the stream with .end()

digitalsadhu commented on lib/template.js:
```diff
@@ -0,0 +1,86 @@
+        body.send(documentTrailers);
+        body.end();
+    });
+};
```
Not much difference here between this and the original built in html template. We check if body is a response stream instead of a string and if it is, we set up streaming, otherwise we do the same as before and just spit out a template

digitalsadhu commented on example/streaming/server.js:
```diff
@@ -0,0 +1,169 @@
+    const contentFetch = content.fetch(incoming);
+    const footerFetch = footer.fetch(incoming);
+
+    incoming.hints.on('complete', async ({ js, css }) => {
```
once we have hints in from the podlets, we can set up streaming

digitalsadhu commented on example/streaming/server.js:
```diff
@@ -0,0 +1,169 @@
+    incoming.hints.on('complete', async ({ js, css }) => {
+        // set the assets on httpincoming so that they are available in the document template
+        incoming.js = [...incoming.js, ...js];
+        incoming.css = [...incoming.css, ...css];
```
this is pretty awkward but works for now at least

digitalsadhu commented on example/streaming/server.js:
```diff
@@ -0,0 +1,169 @@
+                    </div>
+                </div>
+            </template>
+        `);
```
In this case we are sending a DSD template with slots and placeholder skeleton html so that we can replace these by streaming in podlet content later.

digitalsadhu commented on example/streaming/server.js:
```diff
@@ -0,0 +1,169 @@
+
+        // close out the dom and the stream
+        await Promise.all([headerFetch, menuFetch, contentFetch, footerFetch]);
+        stream.done();
```
Indicate that we are finished streaming content so that the html template can finish things off

digitalsadhu commented on example/streaming/server.js:
```diff
@@ -0,0 +1,169 @@
+        // stream in podlet content when available...
+        headerFetch.then((content) => {
+            stream.send(`<div slot="header">${content}</div>`);
+        });
```
we decorate the podlet content in divs with slot attributes to indicate where they should be placed into the document.

wkillerud commented on lib/layout.js:
```diff
@@ -615,6 +618,34 @@ export default class PodiumLayout {
+
+                    // call our document template, injecting a token we can use to split the template
+                    // into header and footer
+                    const splitToken = `<!--split-token-->`;
```
I'm having a tough time visualizing how this split would affect loading. Is this only here to handle the `server-simple` case of shipping off `<head>` and `<body>` in two chunks?

digitalsadhu replied:
We split the template on the position that would normally have the body injected by the developers so that we are left with 2 pieces that we are able to deliver. The first bit goes immediately, then we allow the user to stream whatever they want, and then we deliver the second bit at the end. In practice this looks like...

We deliver this right away..

```js
<html>
<head>
...
</head>
<body>
```

Then the developer does there thing and streams the content they want in the body... podlets etc

```js
<main>
... podlets etc...
</main>
```

Then we deliver the second split part of the template which is the document end...

```js
... some scripts etc...
</body>
</html>
```

digitalsadhu replied:
The nice thing about using the split technique is that all existing templates should just work as they are so its nice and backwards compatible

wkillerud replied:
I guess my main concern about shipping a "finished" `<head>` is that podlets' early hints might give us styles and scripts that should be up there. I suppose that's a design constraint with streaming you have to deal with as an app developer, that your stuff can come in at any time?

digitalsadhu replied:
we wait for all assets before shipping the head by using incoming.hints.on('complete')

wkillerud reviewed (commented):
This looks very cool 🎉

digitalsadhu commented:
Looked into browser and server support and the need for fallbacks etc. As I understand it, we are good because:
1. HTTP 1.1 onwards supports streams (HTTP 2 is more efficient)
2. If a browser or server didn't support streaming, content would just be buffered and displayed at once
3. Early hints work between podlet and layout regardless because we control the client (Undici) which supports early hints
4. Early hints won't work between the layout and the browser for older browsers or servers not using HTTP 2
5. Again, we dont need a fallback for early hints between layout and browser, all it will mean is a perf degradation.

trygve-lie commented:
This is great stuff!

On the general side; are the loading sequential or are they out of order? In the video they look sequential. Getting it rendered out of order is a bit different problem I think.

digitalsadhu commented:
> This is great stuff!
> 
> On the general side; are the loading sequential or are they out of order? In the video they look sequential. Getting it rendered out of order is a bit different problem I think.

You have full control, in my demo I just did sequential setTimeouts to show the delay. You can have it all come in when it comes in or you can use Promise.all to wait for all podlets and send them out at once or you could do the header first and the rest after or however you like.

digitalsadhu commented:
I've reworked streaming after the loss of early hints. Link headers are still used between the podlets and the layout in order to get assets to the layout ahead of the rest of the podlet content.


https://github.com/user-attachments/assets/60f0912e-3163-43c0-912e-ae3e712c2c63


https://github.com/user-attachments/assets/3619c42b-fa3b-46b4-91de-fb05f1223968


I've changed the podiumStream() method to return a promise and to wait for podlet assets to be ready and then flush the document head in the background to simplify the API surface.

Take a look at the examples folder for several examples on how this works

leftieFriele commented:
`.sendHeaders` is something you would put in a hook in servers like Fastify, and I assume we could do something similar for Express where we perform `.sendHeaders` as part of the middleware in `podlet`? without developers having to remember to add it.

digitalsadhu commented:
> `.sendHeaders` is something you would put in a hook in servers like Fastify, and I assume we could do something similar for Express where we perform `.sendHeaders` as part of the middleware in `podlet`? without developers having to remember to add it.

The trouble is that if we do, the developer cannot modify headers at all. If they do, headers will be double sent which will throw an error. This is why early hints was so good. Unfortunately we need to have the developers call .sendHeaders in case they need to set headers themselves

wkillerud commented on README.md:
```diff
@@ -613,6 +613,44 @@ app.get(layout.pathname(), (req, res) => {
+
+The Podium document template will still be used. When you call res.podiumStream(), the document head will be sent to the browser immediately. Once the .done() function is called, the closing part of the template will be sent before the stream is closed out.
+
+Note that any arguments passed to .podiumStream(...args) will be passed on to the layout's document template.
```
I wonder if this should be an options object so we can add to it without causing a breaking change.

```
res.podiumStream({
  view: {
    templateArgs: [...args],
  },
});
```

digitalsadhu replied:
I don't think there are any breaking changes to worry about here. Did you spot something?

wkillerud replied:
Not now, I'm thinking if we want to add stuff in the future.

wkillerud commented on README.md:
````diff
@@ -613,6 +613,44 @@ app.get(layout.pathname(), (req, res) => {
+
+**Working with assets**
+
+When working with assets, its important to wait for podlets to have sent their assets to the layout via 103 early hints. Use the incoming.hints `complete` event for this. Wait for assets to be ready, set assets on the incoming object and then call res.podiumStream.
+
+```js
+const incoming = res.locals.podium;
+const headerFetch = p1Client.fetch(incoming);
+const footerFetch = p2Client.fetch(incoming);
+
+incoming.hints.on('complete', async ({ js, css }) => {
+    incoming.js = js;
+    incoming.css = css;
+
+    const stream = res.podiumStream();
+    const [header, footer] = await Promise.all([
+        headerFetch,
+        footerFetch,
+    ]);
+    stream.send(`<header>${header}</header>...<footer>${footer}</footer>`);
+    stream.done();
+});
+```
````
Is this bit still correct? I don't see it in the examples directory.

digitalsadhu replied:
needs updating, still talks about 103 early hints

wkillerud commented on example/streaming/podlets/content.js:
```diff
@@ -0,0 +1,61 @@
+});
+
+app.get('/', async (req, res) => {
+    res.set('Content-Type', 'text/html');
+    res.sendHeaders();
+
+    await new Promise((res) => setTimeout(res, 2200));
+
+    res.podiumSend(`
+			<section class="content">
+				<h1>Podlets fetched and composed, on demand, just for you</h1>
+				<p>Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam. Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam.</p>
+				<p>Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam. Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam.</p>
+			</section>	
+		`);
```
The layout examples have some inline documentation explaining the different parts. It would be nice to have something similar for at least one of the podlets, or in a separate README.

If it's covered in the podlet-repo, perhaps link to docs there?
